SceneUtils: Add reduceVertices().

elalish opened the pull request:
I wrote this some time ago to support model-viewer's automatic camera framing (which is tight without allowing the object to leave the frame for any orbit parameters with constant zoom). I always thought it  might be a useful generic method for three.js, so I figured I'd contribute it and  see if anyone agrees it might be useful to others.

Mugen87 commented:
`reduceVertices()` works with the `geometry` which is no property of `Object3D`. That makes me feel that `Object3D` is not the right place for this method.

For the time begin I would suggest to add the method to [SceneUtils](https://github.com/mrdoob/three.js/blob/dev/examples/jsm/utils/SceneUtils.js).

elalish commented:
Sorry for the delay, I've addressed the comments now. I was reminded of this while fixing this function in MV for skinned meshes: https://github.com/google/model-viewer/pull/3272

I was hoping to fix three's bounding box computation, but since it's inside of BufferGeometry it can't access the bones. This function can be a work-around in the meanwhile, but it seems like there should be a more core way of handling this.

Mugen87 commented:
> I was hoping to fix three's bounding box computation, but since it's inside of BufferGeometry it can't access the bones.

Related #21507

elalish commented:
Okay, updated from `traverse` to `traverseVisible` to allow some control. Also ran `build-examples` to generate the js version, which conveniently told me about some includes I'd forgotten. Should be ready now @mrdoob.

elalish commented:
Actually, @donmccurdy would you mind taking a look at this? It seems to work quite well generally, but it still seems to fail your example: https://github.com/google/model-viewer/issues/2450#issuecomment-902428164

elalish commented:
Interesting, so think I found the final remaining problem. The issue with the old CesiumMan model was that is used `unorm8` for its skin weights vector. That's handled properly in the shaders, but in SkinnedMesh -> boneTransform it is not. I tried adding `getNormalizedComponentScale( geometry.attributes.skinWeight.array.constructor );` to that function, but three does not like importing a function from an example. @Mugen87 @donmccurdy any thoughts on the right way to solve this? Maybe `getNormalizedComponentScale` should be a member of `BufferAttribute`?

donmccurdy commented:
It does seem like we're going to need something like https://github.com/mrdoob/three.js/pull/22874 sooner or later, either directly on THREE.BufferAttribute or in a utility function.

elalish commented:
Agreed; is there any problem with #22874? I'd like to merge that first so I can clean this up and test it.

donmccurdy commented:
Just bumping this thread now that https://github.com/mrdoob/three.js/pull/22874 has been merged. :)

elalish committed:
updated

elalish commented:
Okay, I think this one's ready now. Supposedly I should run `npm run build-examples` to update the `js` directory, but when I do, I get a pile of unrelated changes. What should I do?

LeviPesin commented:
> Supposedly I should run npm run build-examples to update the js directory

You shouldn't, all building is done by mrdoob in separate commits.

Mugen87 commented on examples/jsm/utils/SceneUtils.js:
```diff
@@ -118,8 +118,55 @@ function createMultiMaterialObject( geometry, materials ) {
+	let value = initialValue;
+	const vertex = new Vector3();
+
+	this.traverseVisible( ( object ) => {
```
Since `reduceVertices()` is no method of `Object3D` anymore, `this` does not point to the right spot here. I guess the function needs an additional parameter representing the 3D object.

elalish replied:
Oh, good call. I was also asked to add this to an example; I'll do that too to make sure I don't miss any other errors.

Mugen87 replied:
I've resolved the merge conflicts and updated `reduceVertices()` so it works now in context of `SceneUtils`. The function has now three instead of two parameters.

Mugen87 committed:
Merge branch 'dev' into reduce

Mugen87 committed:
Update SceneUtils.js

Mugen87 committed:
Update SceneUtils.js

Fix import.

elalish commented:
Thank you!

OndrejSpanel commented:
> Sort of, but the difference is that it does not have to reduce to a `Vertex3` type. For instance, you could reduce to a field of view for a given radius or something of that nature. Since the type is unknown, you have to supply an initial value, because there is no default constructor.

Note: In functional languages such function is usually called `fold`, not `reduce`.
